src/module_handle.cc

```diff
@@ -0,0 +1,279 @@
+  const std::size_t length = this->_module->Deref()->GetModuleRequestsLength();
+  Isolate* isolate = Isolate::GetCurrent();
+  return v8::Integer::New(isolate, length);
+}
+
+template <int async>
+v8::Local<v8::Value> ModuleHandle::GetModuleRequest(v8::Local<v8::Value> index) {
+	return ThreePhaseTask::Run<async, GetModuleRequestRunner>(*this->isolate, this->_module, index->Int32Value());
+}
+
+
+v8::Local<v8::Value> ModuleHandle::SetDependency(v8::Local<v8::Value> value, ModuleHandle* module_handle_ptr) {
+	if (!value->IsString()) {
+		return v8::Boolean::New(v8::Isolate::GetCurrent(), false);
+	}
+	v8::Local<v8::String> key = value->ToString();
+	v8::String::Utf8Value keyAsCharPointer(v8::Isolate::GetCurrent(), key);
+	if (!keyAsCharPointer.length()) {
+		// UTF8-conversion error probably
+		return v8::Boolean::New(v8::Isolate::GetCurrent(), false);
+	}
+	// Assume the code cannot invoke SetDependency() in parallel, otherwise will the code below not be thread safe
+	this->dependencies->insert(dependency_map_type::value_type(*keyAsCharPointer, module_handle_ptr->_module));
+	return v8::Boolean::New(v8::Isolate::GetCurrent(), true);
+}
+
+
+std::mutex InstantiateRunnerMutex;
+std::shared_ptr<ModuleHandle::dependency_map_type> InstantiateRunnerDependencies;
+
+struct InstantiateRunner : public ThreePhaseTask {
+	shared_ptr<RemoteHandle<Context>> context;
+	std::shared_ptr<RemoteHandle<v8::Module>> _module;
+	std::unique_ptr<Transferable> result;
+	std::lock_guard<std::mutex> lock_guard;
+	
+	
+	InstantiateRunner(IsolateHolder* isolate, ContextHandle* context_handle, std::shared_ptr<RemoteHandle<v8::Module>> myModule, std::shared_ptr<ModuleHandle::dependency_map_type> deps)
+		: context(context_handle->context), _module(std::move(myModule)), lock_guard(InstantiateRunnerMutex)
+	{
+		// Sanity check
+		context_handle->CheckDisposed();
+		if (isolate != context_handle->context->GetIsolateHolder()) {
+			throw js_generic_error("Invalid context");
+		}
+		InstantiateRunnerDependencies = std::move(deps);
+	}
+
+	static v8::MaybeLocal<v8::Module> ResolveCallback(v8::Local<v8::Context> context, v8::Local<v8::String> specifierAsLocal, v8::Local<v8::Module> referrer) {
+		v8::String::Utf8Value specifierHelper(v8::Isolate::GetCurrent(), specifierAsLocal);
+		const std::size_t length = specifierHelper.length();
+		if (length) {
+			std::string specifierAsString(*specifierHelper, length);
+			// We can safely use InstantiateRunnerDependencies
+			auto & dependencies = InstantiateRunnerDependencies;
+			ModuleHandle::dependency_map_type::iterator it = dependencies->find(specifierAsString);
+			if (it != dependencies->end()) {
+				return v8::MaybeLocal<v8::Module>(it->second->Deref());
+			}
+		}
+		return v8::MaybeLocal<v8::Module>();
+	}
+
+	void Phase2() final {
+		v8::Isolate* isolate = v8::Isolate::GetCurrent();
+		v8::Maybe<bool> maybe = this->_module->Deref()->InstantiateModule(this->context->Deref(), InstantiateRunner::ResolveCallback);
```

Comment:
Just call `Unmaybe()` on this, that will propagate any v8 exceptions. The below line can be removed in that case, and Phase3() would be empty.